lib/api/src/externals/function.rs

```diff
@@ -67,6 +67,49 @@ pub struct Function {
     pub(crate) exported: ExportFunction,
 }
 
+fn build_export_function_metadata<Env>(
+    env: Env,
+    import_init_function_ptr: for<'a> fn(
+        &'a mut Env,
+        &'a crate::Instance,
+    ) -> Result<(), crate::HostEnvInitError>,
+) -> (*mut std::ffi::c_void, ExportFunctionMetadata)
+where
+    Env: Clone + Sized + 'static + Send + Sync,
+{
+    let import_init_function_ptr = Some(unsafe {
+        std::mem::transmute::<fn(_, _) -> Result<(), _>, fn(_, _) -> Result<(), _>>(
+            import_init_function_ptr,
+        )
+    });
+    let host_env_clone_fn: fn(*mut std::ffi::c_void) -> *mut std::ffi::c_void = |ptr| {
+        let duped_env: Env = unsafe {
+            let ptr: *mut Env = ptr as _;
+            let item: &Env = &*ptr;
+            item.clone()
+        };
+        Box::into_raw(Box::new(duped_env)) as _
+    };
+    let host_env_drop_fn: fn(*mut std::ffi::c_void) = |ptr: *mut std::ffi::c_void| {
+        unsafe { Box::from_raw(ptr as *mut Env) };
+    };
+    let env = Box::into_raw(Box::new(env)) as _;
+
+    // # Safety
+    // - All these functions work on all threads
+    // - The host env is `Send`.
+    let metadata = unsafe {
+        ExportFunctionMetadata::new(
+            env,
+            import_init_function_ptr,
+            host_env_clone_fn,
+            host_env_drop_fn,
+        )
+    };
+
+    (env, metadata)
```

Comment:
Why not just returning `metadata` here, and let the rest of the code accessing `metadata.env` to get the environment? Is it by convenience?

Reply:
`metadata.env` is private! It can only be seen in `wasmer_engine` not in `wasmer_API`, this is for the best as the contents of the `metadata` struct can be dangerous